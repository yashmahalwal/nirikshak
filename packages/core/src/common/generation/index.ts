export { generateFaker } from "./fakerGen";
export { generateCustom } from "./customGen";
