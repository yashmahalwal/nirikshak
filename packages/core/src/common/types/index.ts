export { SchemaHelpers } from "./helpers";
export { isCustomFunction, CustomFunction } from "./custom";
export { isFakerType } from "./fakerTypes";
