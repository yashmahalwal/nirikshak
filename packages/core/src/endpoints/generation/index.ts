export { generateURL } from "./urlStringGen";
export { getFromResource } from "./resourceStringGen";
export * from "./helpers";
export * from "./input";
