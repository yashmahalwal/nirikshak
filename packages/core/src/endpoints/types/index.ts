export {
    MethodType,
    HeaderMap,
    Cases,
    BodyType,
    BodyInstance,
} from "./helpers";
export { isURLString } from "./urlString";
export { isResourceString } from "./resourceString";
export { isDescription, Description } from "./description";
export { HeaderAndQuery, Inputs } from "./input";
export { HeaderAndStatus, Outputs } from "./output";
