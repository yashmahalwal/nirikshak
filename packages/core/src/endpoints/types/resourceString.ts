export type ResourceString = string;

export function isResourceString(str: any): str is ResourceString {
    return typeof str === "string" && str.startsWith("resource:");
}
