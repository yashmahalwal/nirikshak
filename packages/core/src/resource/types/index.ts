import {
    isBaseType,
    isWithModifiersBaseType,
    isWithModifiersResource,
    isOneOfEntries,
    BaseType,
    WithModifiers,
    OneOfEntries,
} from "./helper";

// type guard: ResourceBase
export function isResourceBase(input: any): input is ResourceBase {
    if (!input || typeof input != "object" || Array.isArray(input))
        return false;
    return Object.keys(input).every(
        (key) =>
            isBaseType(input[key]) ||
            isWithModifiersBaseType(input[key]) ||
            isWithModifiersResource(input[key]) ||
            isOneOfEntries(input[key]) ||
            isResourceBase(input[key])
    );
}

// A basic key value pair
export interface ResourceBase {
    [key: string]:
        | BaseType
        | WithModifiers<BaseType>
        | WithModifiers<ResourceBase>
        | OneOfEntries
        // Check for resource base at last as it matches any key
        | ResourceBase;
}

// A resource schema is a resource base along with an identifier field
export interface Resource extends ResourceBase {
    id: Exclude<BaseType, Array<any> | null | boolean>;
}

// type guard: isResource
export function isResource(input: any): input is Resource {
    if (!input || typeof input !== "object") return false;

    return (
        !Array.isArray(input["id"]) &&
        isBaseType(input.id) &&
        isResourceBase(input)
    );
}
