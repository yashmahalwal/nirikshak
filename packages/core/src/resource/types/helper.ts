import {
    CustomFunctionType,
    isCustomFunction,
} from "../../common/types/custom";
import { Literal, isLiteral } from "../../common/types/literals";
import { FakerType, isFakerType } from "../../common/types/fakerTypes";
import { ResourceBase, isResourceBase } from "./resource";
import { Primitives } from "../../common/types/helpers";

// An instance is of the following type
export interface ResourceInstanceBase {
    [key: string]: Primitives | ResourceInstanceBase | ResourceInstanceBase[];
}

export interface ResourceInstance extends ResourceInstanceBase {
    id: string | number;
}

// Base type: Simple literals that a resource field can have
export type BaseType =
    | FakerType
    | CustomFunctionType
    | Literal
    | Array<BaseType>;

// Validation for the same
export function isBaseType(input: any): input is BaseType {
    if (Array.isArray(input)) return input.every((entry) => isBaseType(entry));

    // Ordering is important
    return isFakerType(input) || isCustomFunction(input) || isLiteral(input);
}

// Higher order type: Adding modifiers to a resource field entry
// A resource base field can be: Base Type or another Resource base
// You can add nullable, plural or optional here
// Ex: {type: "faker:random.words", nullable: true}
// Ex: {field: {name: "custom:myName"}, plural: true}
export type WithModifiers<
    T extends BaseType | ResourceBase
> = (T extends BaseType
    ? {
          type: BaseType;
      }
    : {
          field: ResourceBase;
      }) & { plural?: boolean; nullable?: boolean; optional?: boolean };

// Checking the validity of the modifier fields of the HOT
function checkModifiers(
    input: WithModifiers<ResourceBase | BaseType>
): boolean {
    let value = true;
    void (
        "nullable" in input &&
        (value = value && typeof input["nullable"] === "boolean")
    );
    void (
        "optional" in input &&
        (value = value && typeof input["optional"] === "boolean")
    );
    void (
        "plural" in input &&
        (value = value && typeof input["plural"] === "boolean")
    );

    return value;
}

// type guard: WithModifiers<ResourceBase>
export function isWithModifiersResource(
    input: any
): input is WithModifiers<ResourceBase> {
    if (!input || typeof input != "object") return false;

    if (!checkModifiers(input)) return false;

    if ("field" in input) {
        return isResourceBase(input["field"]);
    }
    return false;
}

// type guard: WithModifies<BaseType>
export function isWithModifiersBaseType(
    input: any
): input is WithModifiers<BaseType> {
    if (!input || typeof input != "object") return false;

    if (!checkModifiers(input)) return false;

    if ("type" in input) {
        return isBaseType(input["type"]);
    }
    return false;
}

// type guard: WithModifiers<BaseType> or WithModifiers<ResourceBase>
// Infers from input
export function isWithModifiers(
    input: any
): input is WithModifiers<BaseType> | WithModifiers<ResourceBase> {
    return isWithModifiersBaseType(input) || isWithModifiersResource(input);
}

// Selecting one from multiple choices
export type OneOfEntries =
    | {
          types: Array<BaseType | WithModifiers<BaseType>>;
      }
    | {
          fields: Array<ResourceBase | WithModifiers<ResourceBase>>;
      }
    | {
          fields: Array<ResourceBase | WithModifiers<ResourceBase>>;
          types: Array<BaseType | WithModifiers<BaseType>>;
      };

//   type guard: OneOfEntries
export function isOneOfEntries(input: any): input is OneOfEntries {
    if (!input || typeof input !== "object") return false;
    if (!Array.isArray(input["fields"]) && !Array.isArray(input["types"]))
        return false;

    if (Array.isArray(input["fields"]) && !input["fields"].length) return false;
    if (Array.isArray(input["types"]) && !input["types"].length) return false;

    let value = true;
    value =
        value &&
        (Array.isArray(input["types"])
            ? input["types"].every(
                  (entry) => isBaseType(entry) || isWithModifiersBaseType(entry)
              )
            : true);

    value =
        value &&
        (Array.isArray(input["fields"])
            ? input["fields"].every(
                  (entry) =>
                      isResourceBase(entry) || isWithModifiersResource(entry)
              )
            : true);

    return value;
}
