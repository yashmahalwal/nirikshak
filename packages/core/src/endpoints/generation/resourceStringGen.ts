import { ResourceString } from "../types/resourceString";
import { ResourceInstance} from "../../resource/types/helper";
import _ from "lodash";
import { Primitives } from "../../common/types/helpers";
export function getFromResource(
    resourceString: ResourceString,
    instance: ResourceInstance
):
    | Primitives
    | Omit<ResourceInstance, "id">
    | Array<Omit<ResourceInstance, "id">> {
    const path = resourceString.slice(9);
    if (!_.has(instance, path))
        throw new Error(`Invalid resource instance path: ${path}`);
    return _.get(instance, resourceString.slice(9));
}
