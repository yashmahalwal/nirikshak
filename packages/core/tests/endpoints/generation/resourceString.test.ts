import { ResourceInstance } from "../../../src/resource/types/helper";
import { ResourceString } from "../../../src/endpoints/types/resourceString";
import faker from "faker";
import { getFromResource } from "../../../src/endpoints/generation/resourceStringGen";

const resource = {
    id: faker.random.uuid(),
    name: {
        firstName: faker.name.firstName(),
        lastName: faker.name.lastName(),
    },
    addresses: [
        { zipCode: faker.address.zipCode(), city: faker.address.city() },
        { zipCode: faker.address.zipCode(), city: faker.address.city() },
    ],
    recursion: {
        bases: {
            args: {
                plethora: [1, 2, 10, [[7, 5, [4]]], 3],
            },
        },
    },
};

const ValidInputs: {
    input: ResourceString;
    output: () => ResourceInstance[string];
}[] = [
    {
        input: "resource:id",
        output(): string {
            return resource.id;
        },
    },
    {
        input: "resource:name.lastName",
        output(): string {
            return resource.name.lastName;
        },
    },
    {
        input: "resource:addresses[1].city",
        output(): string {
            return resource.addresses[1].city;
        },
    },
    {
        input: "resource:name",
        output(): typeof resource.name {
            return resource.name;
        },
    },
    {
        input: "resource:recursion.bases.args.plethora[3][0][2][0]",
        output(): number {
            return resource.recursion.bases.args.plethora[3][0][2][0];
        },
    },
];

const InvalidInputs: ResourceString[] = [
    "resource:age",
    "resource:addresses[4].city",
    "resource:recursion.age",
    "resource:recursion.bases.args.plethora[4][0][2][0]",
];

void ValidInputs;

describe("Getting data from resource string", () => {
    test.each(ValidInputs)(`Valid input: %#`, (entry) =>
        expect(getFromResource(entry.input, resource)).toEqual(entry.output())
    );

    test.each(InvalidInputs)(`Invalid input: %#`, (entry) =>
        expect(() =>
            getFromResource(entry, resource)
        ).toThrowErrorMatchingSnapshot()
    );
});
