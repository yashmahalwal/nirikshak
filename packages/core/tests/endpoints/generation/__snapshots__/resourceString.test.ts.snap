// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Getting data from resource string Invalid input: 0 1`] = `"Invalid resource instance path: age"`;

exports[`Getting data from resource string Invalid input: 1 1`] = `"Invalid resource instance path: addresses[4].city"`;

exports[`Getting data from resource string Invalid input: 2 1`] = `"Invalid resource instance path: recursion.age"`;

exports[`Getting data from resource string Invalid input: 3 1`] = `"Invalid resource instance path: recursion.bases.args.plethora[4][0][2][0]"`;
